event.go

```diff
@@ -308,9 +439,23 @@ func (e Event) SetUnsigned(unsigned interface{}) (Event, error) {
 	if eventJSON, err = CanonicalJSON(eventJSON); err != nil {
 		return Event{}, err
 	}
-	result := e
+	result := *e
 	result.eventJSON = eventJSON
-	result.fields.Unsigned = unsignedJSON
+	eventIDFormat, err := e.roomVersion.EventIDFormat()
+	if err != nil {
+		return Event{}, err
+	}
+	if eventIDFormat == EventIDFormatV1 {
+		fields := result.fields.(eventFormatV1Fields)
+		fields.Unsigned = unsignedJSON
+		fields.fixNilSlices()
+		result.fields = fields
+	} else {
+		fields := result.fields.(eventFormatV2Fields)
```

Comment:
It feels odd to be setting v2 without checking that the event ID format is indeed v2? No commentary to explain why this is the right thing to do. Also, you can factor out:
```
		fields.Unsigned = unsignedJSON
		fields.fixNilSlices()
		result.fields = fields
```